Cross compiling fixes and improvements

eigendude opened the pull request:
## Description

This PR contains a set of fixes and improvements to the build system to allow for cross-compiling, and then applies several code fixes needed for newer GCC versions.

### Structure of the PR

The PR is organized as follows:

1. Improvements needed for cross-compiling
2. Build fixes

### Improvements for cross-compiling

Commits:

- [CMake: Expose CMake options to includes](https://github.com/eigendude/libp2p/commit/cross-compile~5)
- [CMake: Exclude test dependencies when building tests](https://github.com/eigendude/libp2p/commit/cross-compile~4)
- [CMake: Allow build system to provide protobuf compiler](https://github.com/eigendude/libp2p/commit/cross-compile~3)

These changes enable more control over the build process when embedding in another build system.

CMake is controlled by user variables; the first one we need is `-DHUNTER_ENABLED=OFF` so that we can have control over dependencies.

Another useful variable is `-DTESTING=OFF`, because embedding test infra drastically increases the up-front effort. The first commit above allows this variable to exclude test infra from the dependency discovery in addition to the build.

Finally, I added two variables to allow the build system to provide the protobuf compiler and headers:

* `-DProtobuf_PROTOC_EXECUTABLE=$(NATIVEPREFIX)/bin/protoc`
* `-DProtobuf_INCLUDE_DIR=$(PREFIX)/include`

We might also want to disable `-Werror` for examples, but I haven't done that here.

### Build fixes

Commits:

- [Fix build error due to missing include](https://github.com/eigendude/libp2p/commit/cross-compile~2)
- [Fix build error due to mismatched types](https://github.com/eigendude/libp2p/commit/cross-compile~1)
- [Fix compiler warning due to sign comparison](https://github.com/eigendude/libp2p/commit/cross-compile)

These fixes should be evident by visual inspection. Error log included in the commit messages.

## How has this been tested?

With these changes in place, I'm able to compile and link libp2p and dependencies using two build systems: OpenEmbedded, and Kodi's depends system.

Here's my work on these two build systems. Commits are structured so that libp2p inclusion can be upstreamed.

### Additions for OpenEmbedded (meta-cryptocurrency layer)

Commits:

* [libp2p: Import C++ implementation of libp2p](https://github.com/eigendude/meta-cryptocurrency/commit/filecoin~1)

This commit enables libp2p support in the OpenEmbedded ecosystem.

### Additions for Kodi

Commits for C++17 support:

* [[docs] Update Linux readmes for C++17 requirement](https://github.com/eigendude/kodi/commit/filecoin~7)
* [[depends] Check for C++17 support](https://github.com/eigendude/kodi/commit/filecoin~6)
* [Change to C++17 for core](https://github.com/eigendude/kodi/commit/filecoin~5)

Kodi currently uses C++14. These 3 commits can be PR'd to enable C++17 support, needed for libp2p and filecoin. Merging is probably a long way off, as Kodi supports a long tail of devices with embedded build systems that often use old compilers.

Commits for the depends build system:

* [tools/depends: Update OpenSSL to v1.1.1](https://github.com/eigendude/kodi/commit/filecoin~4)
* [tools/depends: Add libp2p and dependencies](https://github.com/eigendude/kodi/commit/filecoin~3)

With the commits in this PR, I was able to cross-compile libp2p for x86_64 Linux and macOS for runtime testing. Next is to test compilation for Raspberry Pi and Android NDK.

Commits for runtime testing:

* [testing libp2p: echo, kademlia and gossip](https://github.com/eigendude/kodi/commit/filecoin)

I didn't spend a lot of time on kademlia, but echo and gossip test code produces promising output. Tested on Linux x86_64 and macOS.

## Benefits

IoT is the obvious goal, but Kodi opens up another little-known area of embedded: OTT (over-the-top), stretching across both set-top boxes and almost all mobile

## Related PRs

This PR is needed for https://github.com/filecoin-project/cpp-filecoin/pull/114.

igor-egorov commented on src/common/literals.cpp:
```diff
@@ -11,16 +11,18 @@
   libp2p::common::Hash256 operator""_hash256(const char *c, size_t s) {
     libp2p::common::Hash256 hash{};
-    std::copy_n(c, std::min(s, 32ul), hash.rbegin());
+    std::copy_n(c, std::min(s, static_cast<size_t>(32ul)), hash.rbegin());
```
just a minor note.
It looks like `l` type specifier for constants can be omitted due to explicit `static_cast`.

eigendude replied:
Should I drop the `u` type specifier too?

igor-egorov replied:
`u` could still be kept as an indicator of the value's "nature".
That is not an issue at all. We may leave it as is even with `l`.

eigendude replied:
I removed the `l` and left the `u`. this leaves the indicator of its "nature", and de-duplicates the notion of its "size".

masterjedy replied:
imho, casting to the same type explicitly looks more clear.

igor-egorov commented on src/protocol/gossip/impl/remote_subscriptions.cpp:
```diff
@@ -10,6 +10,9 @@
 #include <libp2p/protocol/gossip/impl/connectivity.hpp>
 #include <libp2p/protocol/gossip/impl/message_builder.hpp>
 
+#define SPDLOG_TRACE_ON
```
The fact you had to define the macro is looking suspicious <img width="16" alt="fry" src="https://user-images.githubusercontent.com/36847043/76315208-4c95a900-62e9-11ea-8ff7-e2ee57c7f70b.png">

eigendude replied:
not sure if needed, or resulted from grasping at straws...

spd was a headache and I had to comment random log lines just to get libp2p to link. I'll put some more time into this.

igor-egorov replied:
Please check if it works when the macro is not defined.

eigendude replied:
I dropped the commit entirely.

eigendude commented on cmake/functions.cmake:
```diff
@@ -45,8 +45,8 @@ function(add_flag flag)
 function(compile_proto_to_cpp PROTO_LIBRARY_NAME PB_H PB_CC PROTO)
-  get_target_property(Protobuf_INCLUDE_DIR protobuf::libprotobuf INTERFACE_INCLUDE_DIRECTORIES)
-  get_target_property(Protobuf_PROTOC_EXECUTABLE protobuf::protoc IMPORTED_LOCATION_RELEASE)
+  #get_target_property(Protobuf_INCLUDE_DIR protobuf::libprotobuf INTERFACE_INCLUDE_DIRECTORIES)
```
Right, I don't intend to leave commented code - commit is marked [WIP]. They should probably be conditionally enabled, but on what condition? They broke my cross compile attempt, so maybe we switch on cross compiling? I don't think we should expose a CMake option.

Do you know why these lines were added in the first place?

igor-egorov replied:
What I see here - in these two lines we are reading `INTERFACE_INCLUDE_DIRECTORIES` and `IMPORTED_LOCATION_RELEASE` parameters, which have to be set by `find_package(Protobuf ...)` command.

The error from the commit message says - it was not succeeded to find a proto compiler in your environment. Probably you can additionally specify own custom CMake toolchain file which will `set_target_property(...)`. Otherwise, it looks like proto couldn't be compiled without a compiler :)

We could employ [CMAKE_CROSSCOMPILING](https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING.html) var:
```cmake
if (NOT CMAKE_CROSSCOMPILING)
  get_target_property(Protobuf_INCLUDE_DIR protobuf::libprotobuf INTERFACE_INCLUDE_DIRECTORIES)
  get_target_property(Protobuf_PROTOC_EXECUTABLE protobuf::protoc IMPORTED_LOCATION_RELEASE)
endif()
``` 
BUT, the variables still have to be defined in another way.

eigendude replied:
I tackled this the following way:

* [CMake: Allow build system to provide protobuf compiler](https://github.com/eigendude/libp2p/commit/cross-compile~3)

igor-egorov commented:
An update: it seems illegal to remove CONFIG from find_package commands :(

eigendude commented:
Hm, CONFIG requires a cmake config file, but it seems this is provided by hunter, whereas cmake natively provides a find module. how would we find packages without hunter?

eigendude commented:
I think I figured it out. I'm re-approaching without removing CONFIG.

eigendude commented:
I've updated the PR. All CONFIG changes were dropped. Instead, I updated my build system to use the hunterized repos. The down side is that these don't seem to be fully up to date.

I tested cross compiling to Android with clang. Only change needed was a compiler warning fix, which I've added.

eigendude commented:
Any word? I compiled libp2p into an NDK app (kodi) and ran echo and kademlia examples successfully on Android.

masterjedy commented:
> Any word? I compiled libp2p into an NDK app (kodi) and ran echo and kademlia examples successfully on Android.

sounds great

igor-egorov commented:
wow, this pr is still alive :) queued for check

eigendude commented:
yup, the PR lives. I started a new project, embedded Interplanetary Identifiers, to see if an identity solution could solve the problem I'm working on. In the course of implementing IPIDs I rebased this PR and run-time tested it again.

FlorianFranzen commented:
@kamilsa Would be nice if some of these fixes could be used in kagome as well, especially the protobuffer fix.

eigendude commented:
@FlorianFranzen thanks!
